cpp/ql/src/experimental/Security/CWE/CWE-020/NoCheckBeforeUnsafePutUser.ql

```diff
@@ -21,7 +21,7 @@ class WriteAccessCheckMacro extends Macro {
   VariableAccess va;
 
   WriteAccessCheckMacro() {
-    this.getName() = ["user_write_access_begin", "user_access_begin"] and
+    this.getName() = ["user_write_access_begin", "user_access_begin", "access_ok"] and
```

Comment:
According to the documentation for [user_access_begin](https://lgtm.com/projects/g/torvalds/linux/snapshot/5f1f17e546170fa08a34f3d7c300196acad54184/files/arch/x86/include/asm/uaccess.h?sort=name&dir=ASC&mode=heatmap#L442), "you have to not only do the access_ok() checking before using them, but you have to surround them with the user_access_begin/end() pair."  So as I understand this use of `access_ok` is not equivalent?

Reply:
Hello @geoffw0,

While writing the definition logic of WriteAccessCheckMacro, I take [uaccess.h](https://github.com/torvalds/linux/blob/master/include/linux/uaccess.h) as reference, in which all 3 marcos defined

```
#ifndef user_access_begin
#define user_access_begin(ptr,len) access_ok(ptr, len)

#ifndef user_write_access_begin
#define user_write_access_begin user_access_begin
```

So I think they are just the same from a view without cpu architecture inforamtion taken into consideration.

Reply:
> Hello @geoffw0,
> 
> While writing the definition logic of WriteAccessCheckMacro, I take [uaccess.h](https://github.com/torvalds/linux/blob/master/include/linux/uaccess.h) as reference, in which all 3 marcos defined
> 
> ```
> #ifndef user_access_begin
> #define user_access_begin(ptr,len) access_ok(ptr, len)
> 
> #ifndef user_write_access_begin
> #define user_write_access_begin user_access_begin
> ```
> 
> So I think they are just the same from a view without cpu architecture inforamtion taken into consideration.

And In some arch only access_ok is used, for example https://github.com/torvalds/linux/blob/1930a6e739c4b4a654a69164dbe39e554d228915/arch/alpha/kernel/signal.c

Reply:
Ah, that makes sense I think.